### PR TITLE
fix: build xcframeworks to fix linking issue in Playground App

### PR DESCRIPTION
**Pull request by okwasniewski**
**Describe the change**

This PR fixes RNTA linking issues for newer versions of CMake, this PR is still a draft as I needed to patch-package the react-native-test-app. I need to make sure this is the correct solution to solve this issue and contribute the fix back to RNTA.

**Screenshots**

If applicable, add screenshots to help explain your change. Consider using tools like *Giphy Capture* or *ScreenToGif* to capture animated gifs. If you are using physical devices rather than emulators/simulators, consider using tools like *Vysor* or *scrcpy* to project your device's display to your dev machine to allow for screen capture.

**Documentation**

Have you updated the [documentation](../README.md) to reflect your changes? (*Yes* or *N/A*)

**Testing**

Have you [tested](../README.md#testing-in-the-playground-app) the final iteration of your changes? (*Yes* or *N/A* or *I need help testing on [Android|iOS|Windows]*)

### CODE REVIEW

**Comment by okwasniewski** [comment]
Hey, I've opened a PR to RNTA to add support for this feature: https://github.com/microsoft/react-native-test-app/pull/2204

**Comment by CedricGuillemet** [comment]
I'm working on fixing the build so we can continue testing this PR.  https://github.com/BabylonJS/BabylonReactNative/pull/657

**Comment by CedricGuillemet** [comment]
@okwasniewski I just merged the build fix PR. You should be able to update your PR and kick a new build. I'll check its progress

**Review by ryantrem — APPROVED** [review]

**Comment by okwasniewski** [comment]
I'm still encountering some issues with react-native-permissions when launching the app. Im going to go back to this tomorrow, so let's wait with merging 🙏

**Comment by okwasniewski** [comment]
Okay, so I got this to work on the simulator, there was an issue with `react-native-permssions` caused by peer dependency resolution but it should be fine now. 

The only thing left is that the Playground is still not building properly for real device (same issue as before). Going to focus on this one now. 

CleanShot 2024-09-13 at 14 44 04@2x

**Comment by okwasniewski** [comment]
Hey @CedricGuillemet @ryantrem 

I've made it work with xcframeworks. 

So here #658, I've mentioned building a single .xcframework to fix the issue, but unfortunately, this wasn't a viable option. 

I wanted to merge all of the static libs that Babylon currently builds into one and then build an xcframework for each platform and architecture, but there were bunch of linking issues resulting from the merge. 

So I decided to create multiple XCFrameworks for all of our static libraries. 

I've integrated it into gulpfile.js and run it on postinstall of RNTA. (This can be further expanded to build on the CI, but for the sake of keeping things simple in the first iteration I created a local script). 

So now when you run `npm install` we build all of the xcframeworks and link them to RNTA using `vendored_frameworks`. 

So currently, we build for `ios-arm64`, `ios-simulator-arm64` (M1 mac) and `ios-simulator-x86` (Intel mac). 

This can be easily expanded for other platforms like macOS and visionOS later. 

It would be great if you could test out the script locally and let me know what you think.

"Build from source" stays as an option but only for CMake < 3.24 (as stated in README)

File: Modules/@babylonjs/react-native-iosandroid/react-native-babylon.podspec [hunk 1]
```diff
@@ -17,47 +19,73 @@ Pod::Spec.new do |s|
...
+        "CLANG_CXX_LANGUAGE_STANDARD" => "c++17"
+      }
+      s.dependency "React-Codegen"
+      s.dependency "RCT-Folly"
+      s.dependency "RCTRequired"
+      s.dependency "RCTTypeSafety"
+      s.dependency "ReactCommon/turbomodule/core"
+      s.dependency "React-RCTFabric"
+    end
+  end
```
Comment (okwasniewski) [inline Modules/@babylonjs/react-native-iosandroid/react-native-babylon.podspec:89]: fyi: This is required by newer versions of React Native

File: Modules/@babylonjs/react-native-windows/package.json [hunk 2]
```diff
@@ -25,11 +25,11 @@
...
-    "@babylonjs/react-native": "version",
-    "react": ">=17.0.1",
-    "react-native": ">=0.64.0",
-    "react-native-permissions": "^3.0.0",
-    "react-native-windows": ">=0.64.1"
+    "@babylonjs/react-native": "*",
+    "react": "*",
+    "react-native": "*",
+    "react-native-permissions": "*",
+    "react-native-windows": "*"
```
Comment (okwasniewski) [inline Modules/@babylonjs/react-native-windows/package.json:32]: Even if I got RNTA to compile, `react-native-permissions` was erroring on startup, removing exact version of peer dependency fixed the issue. Probably there is other way to this let me know if I should change it

**Comment by okwasniewski** [comment]
Also to address any performance concerns, dynamic linking of xcframeworks add some overhead to the app startup, but for frameworks that only ship static `.a` libraries (our case) Cocoapods just copy the proper slice from the framework. 

There is an additional build script being executed "[CP] Copy XCFrameworks" that copies proper static libraries for current build and only they end up in app binary 

CleanShot 2024-09-20 at 14 54 50@2x

**Comment by okwasniewski** [comment]
Hey @ryantrem @CedricGuillemet @bghgary 

I wanted to check in and ask if you are okay with this approach? 

The build is still failing but I want to ensure I'm on the right track before I spend time fixing it. 

I think that we need to build the XCFrameworks only when BabylonNative version is updated. This should significantly speed up the CI builds

File: Apps/BRNPlayground/package.json [hunk 3]
```diff
@@ -14,7 +14,8 @@
     "start": "react-native start",
     "test": "jest",
     "windows": "react-native run-windows --sln windows/BRNPlayground.sln",
-    "postinstall": "node ./postinstall.js"
+    "postinstall": "node ./postinstall.js",
+    "build-xcframeworks": "cd ../../Package && npx gulp createXCFrameworks"
```
Comment (ryantrem) [inline Apps/BRNPlayground/package.json:18]: Should we run this as part of the postinstall? I think we would still want to be able to just run `npm install` in the directory and then be ready to build locally. If I'm understanding correctly, that would require an extra command as is?

Comment (okwasniewski) [reply Apps/BRNPlayground/package.json:18]: So the issue I saw with this is that even for Windows jobs we were building XCFrameworks which is less than ideal.. Not sure maybe we can scope it differenty.

**Comment by okwasniewski** [comment]
Hey, @ryantrem @CedricGuillemet  I've stripped almost everything from the PR except build and linking of XCFrameworks. 

Let me provide a bit more detail of what is actually happening in `createXCFrameworks` function: 

1. We map over platforms and create static libraries (.a) for each platform

```js
const buildCommand = `xcodebuild -sdk ${platform} -arch ${arch} -configuration Release -project ReactNativeBabylon.xcodeproj -scheme BabylonNative build CODE_SIGNING_ALLOWED=NO BUILD_LIBRARY_FOR_DISTRIBUTION=YES`;
```

2. Once we have all static libraries we merge multi-architecture ones (iphonesimulator-arm64 and iphonesimulator-x86) This allows users to use both Intel and Apple Sillicon Macs
3. We iterate over merged build outputs and create XCFrameworks for each one of our libraries so we get (`libastc-encoder.xcframework`, `libBabylonNative.xcframework`, `libbgfx.xcframework` etc.)
4. Copy xcframeworks to `@babylonjs/react-native-iosandroid/ios/libs/`
5. Link xcframeworks using `vendored_frameworks` in Podfile

File: Package/gulpfile.js [hunk 4]
```diff
@@ -235,6 +233,68 @@ const createIOSUniversalLibs = async () => {
   libs.map(lib => exec(`lipo -create iOS/Build/Release-iphoneos/${lib} iOS/Build/Release-iphonesimulator/${lib} -output ${assemblediOSAndroidDir}/ios/libs/${lib}`));
 };
 
+const createXCFrameworks = async () => {
```
Comment (okwasniewski) [inline Package/gulpfile.js:236]: Make this part of postinstall

File: Modules/@babylonjs/react-native-iosandroid/react-native-babylon.podspec [hunk 5]
```diff
@@ -17,47 +19,73 @@ Pod::Spec.new do |s|
...
-                'NativeXR',
-                'SPIRV',
-                'spirv-cross-core',
-                'spirv-cross-msl',
-                'tinyexr',
-                'UrlLib',
-                'Window',
-                'XMLHttpRequest',
-                'xr'
+  if ENV['BUILD_BABYLON_FROM_SOURCE'] == 'true' then
```
Comment (okwasniewski) [inline Modules/@babylonjs/react-native-iosandroid/react-native-babylon.podspec:22]: Let's remove it

**Comment by okwasniewski** [comment]
Hey @ryantrem @CedricGuillemet @bghgary 

I've applied the changes we talked about yesterday. Let me know if I should address something else

**Review by ryantrem — APPROVED** [review]
Looks good to me!

**Pull request merged**